File: benchmarking/guide-assignment/data-preprocessing/create_fake_sceptre_data.py
#!/usr/bin/env python3
import sys
import pandas as pd
import numpy as np
from pathlib import Path

def main():
    if len(sys.argv) != 2:
        print("Usage: python create_fake_sceptre_data.py <output_directory>")
        sys.exit(1)
    
    output_dir = Path(sys.argv[1])
    output_dir.mkdir(parents=True, exist_ok=True)
    
    print("Creating fake sceptre data...")
    
    # Parameters for simulation
    num_cells = 500
    num_grnas = 100
    num_NTs = 10
    num_targets = 10
    num_responses = 100
    
    # Create gRNA names
    grna_names = [f"gRNA_{i+1}" for i in range(num_grnas)] + [f"NT_{i+1}" for i in range(num_NTs)]
    cell_names = [f"CELL_{i+1:04d}" for i in range(num_cells)]
    
    # 1. Simulate gRNA matrix (gRNAs x cells) following your R code
    np.random.seed(42)  # For reproducibility
    grna_matrix = np.zeros((num_grnas + num_NTs, num_cells))
    
    # Each cell gets one primary gRNA expressed
    grnas_expressed = np.random.choice(num_grnas + num_NTs, num_cells, replace=True)
    
    for i in range(num_cells):
        # Primary gRNA gets high expression
        grna_matrix[grnas_expressed[i], i] = np.random.uniform(10, 15, 1)
        
        # Other gRNAs get background expression (sparse)
        other_grnas = [j for j in range(num_grnas + num_NTs) if j != grnas_expressed[i]]
        background_counts = np.random.uniform(1,8, len(other_grnas))
        background_mask = np.random.binomial(1, 0.5, len(other_grnas))
        for j, grna_idx in enumerate(other_grnas):
            grna_matrix[grna_idx, i] = background_counts[j] * background_mask[j]
    
    grna_df = pd.DataFrame(
        grna_matrix,
        index=grna_names,  # gRNA names as rows
        columns=cell_names  # cell names as columns
    )
    
    grna_output = output_dir / "grna_matrix.csv"
    grna_df.to_csv(grna_output, index_label="grna_id")
    print(f"✓ Saved gRNA matrix: {grna_output} ({grna_df.shape[0]} gRNAs x {grna_df.shape[1]} cells)")
    
    # 2. Create response matrix (genes x cells)
    response_names = [f"response_{i+1}" for i in range(num_responses)]
    response_matrix = np.random.poisson(3, size=(num_responses, num_cells))
    
    response_df = pd.DataFrame(
        response_matrix,
        index=response_names,
        columns=cell_names
    )
    response_output = output_dir / "response_matrix.csv"
    response_df.to_csv(response_output, index_label="response_id")
    print(f"✓ Saved response matrix: {response_output} ({num_responses} genes x {num_cells} cells)")
    
    # 3. Create gRNA target mapping
    grnas_per_target = num_grnas // num_targets
    targets = []
    
    # Assign targeting gRNAs to targets
    for target_idx in range(num_targets):
        for grna_in_target in range(grnas_per_target):
            targets.append(f"target_{target_idx + 1}")
    
    # Add non-targeting gRNAs
    for nt_idx in range(num_NTs):
        targets.append("non-targeting")
    
    grna_target_df = pd.DataFrame({
        'grna_id': grna_names,
        'grna_target': targets
    })
    target_output = output_dir / "grna_target_data_frame.csv"
    grna_target_df.to_csv(target_output, index=False)
    print(f"✓ Saved gRNA targets: {target_output} ({num_grnas} targeting + {num_NTs} non-targeting)")
    
    print(f"\nCreated fake sceptre data with:")
    print(f"- {num_cells} cells")
    print(f"- {num_grnas} targeting gRNAs ({grnas_per_target} per target)")
    print(f"- {num_NTs} non-targeting gRNAs")
    print(f"- {num_targets} targets")
    print(f"- {num_responses} response genes")

if __name__ == "__main__":
    main()

File: benchmarking/guide-assignment/data-preprocessing/create_fake_sceptre_data.R
make_data <- function() {


  # Parameters
  num_cells <- 500
  num_grnas <- 100
  num_NTs <- 10
  num_targets <- 10  # should divide num_grnas
  num_responses <- 100
  num_batches <- 4  # must divide num_cells

  set.seed(42)  # For reproducibility

  # Create gRNA and cell names
  grna_names <- c(paste0("gRNA_", seq_len(num_grnas)),
                  paste0("NT_", seq_len(num_NTs)))
  cell_names <- sprintf("CELL_%04d", seq_len(num_cells))

  # 1. Simulate gRNA matrix (gRNAs x cells)
  grna_matrix <- matrix(0, nrow = num_grnas + num_NTs, ncol = num_cells)

  grnas_expressed <- sample(num_grnas + num_NTs, num_cells, replace = TRUE)

  for (i in seq_len(num_cells)) {
    grna_matrix[grnas_expressed[i], i] <- sample(10:15, 1)  # Primary expression

    background_counts <- sample(1:8, num_grnas + num_NTs - 1, TRUE)
    background_mask <- rbinom(num_grnas + num_NTs - 1, 1, 0.5)

    grna_matrix[-grnas_expressed[i], i] <- background_counts * background_mask
  }

  colnames(grna_matrix) <- cell_names
  rownames(grna_matrix) <- grna_names

  # 2. Simulate response matrix (responses x cells)
  response_names <- paste0("response_", seq_len(num_responses))
  response_matrix <- matrix(rpois(num_responses * num_cells, lambda = 3),
                            nrow = num_responses, ncol = num_cells)
  rownames(response_matrix) <- response_names
  colnames(response_matrix) <- cell_names

  # 3. Create gRNA-to-target mapping
  grnas_per_target <- num_grnas %/% num_targets
  targets <- c(
    rep(paste0("target_", 1:num_targets), each = grnas_per_target),
    rep("non-targeting", num_NTs)
  )


  grna_target_df <- data.frame(
    grna_id = grna_names,
    grna_target = targets,
    stringsAsFactors = FALSE
  )

  # 4. create covariate_data_frame
  covariate_data_frame <- data.frame(
    batch = paste0("batch_", rep(1:num_batches, each = num_cells / num_batches))
  )

  # 4. return
  list(grna_matrix=grna_matrix,
       response_matrix=response_matrix,
       grna_target_data_frame = grna_target_df,
       covariate_data_frame = covariate_data_frame)
}
